File: ds_6_string/L8.cpp

//8 字符串转整数
//别人写的很巧妙，好好学学

#include <iostream>
#include <cstring>
using namespace std;
class Solution {
public:
    int myAtoi(string str) {
        int res = 0;
        int i = 0;
        int flag = 1;
        // 1. 检查空格
        while (str[i] == ' ') { i++; }
        // 2. 检查符号
        if (str[i] == '-') { flag = -1; }
        if (str[i] == '+' || str[i] == '-') { i++; }
        // 3. 计算数字
        while (i < str.size() && isdigit(str[i])) {//相当于只要找到一个"+"或者"-"之后只要其之后的不是数字，这个循环就不会执行了
            int r = str[i] - '0';
            // ------ 4. 处理溢出，这是关键步骤 ------
            if (res > INT_MAX / 10 || (res == INT_MAX / 10 && r > 7)) {
                return flag > 0 ? INT_MAX : INT_MIN;
            }
            // ------------------------------------
            res = res * 10 + r;
            i++;
        }
        return flag > 0 ? res : -res;
    }
};


int main(){
    string str1 = "3.1415";
    string str2 = "+-3000";
    string str3 = "-3103";
    cout << Solution().myAtoi(str1)<<endl;
    cout << Solution().myAtoi(str2)<<endl;
    cout << Solution().myAtoi(str3)<<endl;
    return 0;
}

File: python-algorithm/Big_Company_Recent/bytedance_add1/L8.py
# -*- coding: utf-8 -*-
INT_MAX=2**31-1
INT_MIN=-2**31


# 方法一
class Solution:
    def myAtoi(self, str: str) -> int:
        import re
        mathes=re.match('[ ]*([+-]?\d+)',str)
        if not mathes:
            return 0
        ans=int(mathes.group(1))
        return min(max(ans,-2**31),2**31-1)


# 方法二
class Automata:
    def __init__(self):
        self.state="start"
        self.ans=0
        self.sign=1
        self.table={
            "start":["start","sign","innum","end"],
            "sign":["end","end","innum","end"],
            "innum":["end","end","innum","end"],
            "end":["end","end","end","end"],
        }

    def get_s(self,c):
        if c.isspace():
            return 0
        if c=='+' or c=='-':
            return 1
        if c.isdigit():
            return 2
        return 3

    def get(self,c):
        self.state=self.table[self.state][self.get_s(c)]
        if self.state=="innum":
            self.ans=self.ans*10+int(c)
            self.ans= min(self.ans,INT_MAX) if self.sign==1 else min(self.ans,-INT_MIN)
        if self.state=="sign":
            if c=='-':
                self.sign=-1


class Solution:
    def myAtoi(self, str: str) -> int:
        Auto=Automata()
        for c in str:
            Auto.get(c)
        return Auto.ans*Auto.sign